File: clients/android/app/src/main/kotlin/my/ssdid/drive/presentation/auth/InviteAcceptViewModel.kt
package my.ssdid.drive.presentation.auth

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import my.ssdid.drive.domain.model.TokenInvitation
import my.ssdid.drive.domain.model.TokenInvitationError
import my.ssdid.drive.domain.repository.AuthRepository
import my.ssdid.drive.util.Result
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * UI state for the invitation acceptance screen.
 */
data class InviteAcceptUiState(
    // Invitation info
    val token: String = "",
    val invitation: TokenInvitation? = null,
    val isLoadingInvitation: Boolean = true,
    val invitationError: String? = null,

    // Registration state (via SSDID Wallet)
    val isLoading: Boolean = false,
    val isWaitingForWallet: Boolean = false,
    val isRegistered: Boolean = false,
    val registrationError: String? = null
)

/**
 * ViewModel for handling invitation acceptance via SSDID Wallet.
 *
 * The flow is:
 * 1. Load invitation info from token
 * 2. User taps "Accept with SSDID Wallet"
 * 3. App creates a challenge and launches wallet via deep link
 * 4. Wallet handles DID creation and service registration
 * 5. Wallet calls back with session token
 */
@HiltViewModel
class InviteAcceptViewModel @Inject constructor(
    private val authRepository: AuthRepository,
    private val savedStateHandle: SavedStateHandle
) : ViewModel() {

    private val _uiState = MutableStateFlow(InviteAcceptUiState())
    val uiState: StateFlow<InviteAcceptUiState> = _uiState.asStateFlow()

    init {
        // Get token from navigation arguments
        val token = savedStateHandle.get<String>("token") ?: ""
        // Restore isWaitingForWallet across process death
        val wasWaiting = savedStateHandle.get<Boolean>("isWaitingForWallet") ?: false
        _uiState.update { it.copy(token = token, isWaitingForWallet = wasWaiting) }

        if (token.isNotBlank()) {
            loadInvitationInfo(token)
        } else {
            _uiState.update {
                it.copy(
                    isLoadingInvitation = false,
                    invitationError = "Invalid invitation link"
                )
            }
        }
    }

    /**
     * Load invitation info from token.
     */
    private fun loadInvitationInfo(token: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoadingInvitation = true, invitationError = null) }

            when (val result = authRepository.getInvitationInfo(token)) {
                is Result.Success -> {
                    val invitation = result.data
                    if (invitation.valid) {
                        _uiState.update {
                            it.copy(
                                isLoadingInvitation = false,
                                invitation = invitation
                            )
                        }
                    } else {
                        val errorMessage = when (invitation.errorReason) {
                            TokenInvitationError.EXPIRED -> "This invitation has expired"
                            TokenInvitationError.REVOKED -> "This invitation has been revoked"
                            TokenInvitationError.ALREADY_USED -> "This invitation has already been used"
                            TokenInvitationError.NOT_FOUND -> "Invitation not found"
                            null -> "This invitation is no longer valid"
                        }
                        _uiState.update {
                            it.copy(
                                isLoadingInvitation = false,
                                invitation = invitation,
                                invitationError = errorMessage
                            )
                        }
                    }
                }
                is Result.Error -> {
                    _uiState.update {
                        it.copy(
                            isLoadingInvitation = false,
                            invitationError = result.exception.message ?: "Failed to load invitation"
                        )
                    }
                }
            }
        }
    }

    /**
     * Retry loading invitation info.
     */
    fun retryLoadInvitation() {
        val token = _uiState.value.token
        if (token.isNotBlank()) {
            loadInvitationInfo(token)
        }
    }

    /**
     * Accept the invitation via SSDID Wallet.
     * Launches the wallet with the ssdid://invite deep link — wallet handles
     * email verification and authentication for new users.
     */
    fun acceptWithWallet() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, registrationError = null) }
            try {
                // Launch wallet with invite deep link — wallet handles email verification + authentication
                authRepository.launchWalletInvite(_uiState.value.token)
                _uiState.update { it.copy(isLoading = false, isWaitingForWallet = true) }
                savedStateHandle["isWaitingForWallet"] = true
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(isLoading = false, registrationError = e.message)
                }
            }
        }
    }

    /**
     * Handle the callback from SSDID Wallet with the session token.
     */
    fun handleWalletCallback(sessionToken: String) {
        viewModelScope.launch {
            try {
                authRepository.saveSession(sessionToken)
                savedStateHandle["isWaitingForWallet"] = false
                _uiState.update { it.copy(isWaitingForWallet = false, isRegistered = true) }
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(isWaitingForWallet = false, registrationError = e.message)
                }
            }
        }
    }

    /**
     * Handle an error returned from SSDID Wallet during invitation acceptance.
     */
    fun handleWalletError(message: String) {
        savedStateHandle["isWaitingForWallet"] = false
        _uiState.update { it.copy(isWaitingForWallet = false, registrationError = message) }
    }
}

File: clients/ios/SsdidDrive/SsdidDrive/Presentation/Auth/InviteAcceptViewModel.swift
import Foundation
import Combine

/// Delegate for invite accept view model coordinator events
protocol InviteAcceptViewModelCoordinatorDelegate: AnyObject {
    func inviteAcceptViewModelDidRegister()
    func inviteAcceptViewModelDidRequestLogin()
    func inviteAcceptViewModelDidRequestEmailRegister(token: String)
    func inviteAcceptViewModelDidRequestOidc(provider: String, token: String)
}

/// View model for invitation acceptance screen.
/// Uses wallet-based flow: loads invitation info, launches SSDID Wallet,
/// and handles the wallet callback with a session token.
@MainActor
final class InviteAcceptViewModel: BaseViewModel {

    // MARK: - Published Properties

    @Published var invitation: TokenInvitation?
    @Published var isLoadingInvitation = true
    @Published var invitationError: String?
    @Published var isWaitingForWallet = false
    @Published var isAcceptingAsExisting = false
    @Published var acceptError: String?
    @Published var registrationError: String?

    // MARK: - Properties

    private let authRepository: AuthRepository
    let token: String
    weak var coordinatorDelegate: InviteAcceptViewModelCoordinatorDelegate?

    private var loadTask: Task<Void, Never>?
    private var acceptTask: Task<Void, Never>?
    private var callbackTask: Task<Void, Never>?

    deinit {
        loadTask?.cancel()
        acceptTask?.cancel()
        callbackTask?.cancel()
    }

    // MARK: - Computed Properties

    var email: String { invitation?.email ?? "" }

    // MARK: - Initialization

    init(authRepository: AuthRepository, token: String) {
        self.authRepository = authRepository
        self.token = token
        super.init()

        loadInvitationInfo()
    }

    // MARK: - Actions

    func loadInvitationInfo() {
        loadTask?.cancel()
        isLoadingInvitation = true
        invitationError = nil

        loadTask = Task { [weak self] in
            guard let self else { return }
            do {
                let info = try await authRepository.getInvitationInfo(token: token)
                guard !Task.isCancelled else { return }
                self.invitation = info
                self.isLoadingInvitation = false

                if !info.valid {
                    self.invitationError = info.errorReason?.displayMessage ?? "This invitation is no longer valid"
                }
            } catch {
                guard !Task.isCancelled else { return }
                self.isLoadingInvitation = false
                self.invitationError = error.localizedDescription
            }
        }
    }

    func acceptWithWallet() {
        guard !isWaitingForWallet, !isLoading else { return }
        isLoading = true
        registrationError = nil

        acceptTask?.cancel()
        acceptTask = Task { [weak self] in
            guard let self else { return }
            do {
                try await authRepository.launchWalletInvite(token: token)
                guard !Task.isCancelled else { return }
                isWaitingForWallet = true
                isLoading = false
            } catch {
                guard !Task.isCancelled else { return }
                isLoading = false
                registrationError = error.localizedDescription
            }
        }
    }

    func handleWalletCallback(sessionToken: String) {
        guard isWaitingForWallet, callbackTask == nil else { return }
        callbackTask = Task { [weak self] in
            guard let self else { return }
            do {
                try await authRepository.saveSessionFromWallet(sessionToken: sessionToken)
                guard !Task.isCancelled else { return }
                isWaitingForWallet = false
                coordinatorDelegate?.inviteAcceptViewModelDidRegister()
            } catch {
                guard !Task.isCancelled else { return }
                isWaitingForWallet = false
                registrationError = error.localizedDescription
            }
        }
    }

    func handleWalletError(message: String) {
        isWaitingForWallet = false
        registrationError = message
    }

    /// Accept the invitation as an already-authenticated user.
    /// Uses the invitation token to accept without re-registration.
    func acceptAsExistingUser() {
        guard invitation != nil else { return }
        isAcceptingAsExisting = true
        acceptError = nil

        acceptTask?.cancel()
        acceptTask = Task { [weak self] in
            guard let self else { return }
            do {
                try await authRepository.acceptInvitationAsExistingUser(token: token)
                guard !Task.isCancelled else { return }
                self.isAcceptingAsExisting = false
                self.coordinatorDelegate?.inviteAcceptViewModelDidRegister()
            } catch {
                guard !Task.isCancelled else { return }
                self.isAcceptingAsExisting = false
                self.acceptError = error.localizedDescription
            }
        }
    }

    /// Handle OIDC authentication result for invitation acceptance.
    /// Registers a new user via OIDC provider and accepts the invitation in one step.
    func handleOidcResult(provider: String, idToken: String) {
        isLoading = true
        errorMessage = nil

        acceptTask?.cancel()
        acceptTask = Task { [weak self] in
            guard let self else { return }
            do {
                try await authRepository.acceptInvitationWithOidc(
                    token: token,
                    provider: provider,
                    idToken: idToken
                )
                guard !Task.isCancelled else { return }
                self.isLoading = false
                self.coordinatorDelegate?.inviteAcceptViewModelDidRegister()
            } catch {
                guard !Task.isCancelled else { return }
                self.isLoading = false
                self.handleError(error)
            }
        }
    }

    func requestLogin() {
        coordinatorDelegate?.inviteAcceptViewModelDidRequestLogin()
    }

    func requestEmailRegister() {
        coordinatorDelegate?.inviteAcceptViewModelDidRequestEmailRegister(token: token)
    }

    func requestOidc(provider: String) {
        coordinatorDelegate?.inviteAcceptViewModelDidRequestOidc(provider: provider, token: token)
    }
}
